Handle special kwargs in user-written Triton kernel calls

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122280

Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

aakhundov committed:
Handle special kwargs in user-written Triton kernel calls

Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

[ghstack-poisoned]

aakhundov committed:
Update on "Handle special kwargs in user-written Triton kernel calls"


Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov committed:
Update on "Handle special kwargs in user-written Triton kernel calls"


Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented on torch/_dynamo/variables/functions.py:
```diff
@@ -831,12 +831,36 @@ def __init__(self, kernel, kernel_idx, grid, **kwargs):
+                # new one with a single config with special kwargs
+                config = Config(kwargs={}, **special_kwargs)
+                new_kernel = autotune(configs=[config], key=[])(self.kernel)
+                kernel_side_table.replace_kernel(self.kernel, new_kernel)
```
this is not safe. There might be another thread that is using the kernel. What you should do instead is once you create a new autotune, or append to the existing one, create a new `TritonKernelVariable` similar to how we do for call_method below, and add this as a new entry to the side table.

aakhundov replied:
What does "using" in "another thread that is using the kernel" mean? Does another thread somehow rely on the old representation (without `Autotune` or with `Configs` without `num_warps` etc.)? Or does it mean that another thread is using the kernel id? If it's the latter, as the replacement is more general / should be compatible with the old call structure, would that not be fine?

oulgen replied:
Another thread is using the kernel id. The other thread perhaps passed different num_warps? Or perhaps did not pass num_warps at all. It is not correct to change the kernel.

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -45,6 +45,13 @@ def add_kernel(self, kernel) -> int:
             self.kernel_to_id[kernel] = idx
             return idx
 
+    def replace_kernel(self, old_kernel, new_kernel):
```
so you wont need this

oulgen commented on torch/_dynamo/variables/functions.py:
```diff
@@ -831,12 +831,36 @@ def __init__(self, kernel, kernel_idx, grid, **kwargs):
         from .lists import BaseListVariable
 
+        special_kwargs = {}
+        for name in ("num_warps", "num_stages", "num_ctas"):
```
Looks like we dont actually read num_ctas from triton.config
https://github.com/pytorch/pytorch/blob/e7141d117f5f870dab61e3a991bf05dab61ed220/torch/_inductor/codegen/wrapper.py#L1117-L1124

oulgen replied:
You should also check whether inductor autotuner actually reads the value of num_ctas and does the right thing with it

aakhundov replied:
Inductor doesn't seem to render `num_ctas` currently in the codegen. Still, I think, we should move `num_ctas` to `Config` in this PR, and possibly add rendering `num_ctas` in the Inductor codegen as a follow-up.

aakhundov replied:
For now, raising `Unsupported` if `num_ctas` is passed as a direct kernel argument.

aakhundov committed:
Update on "Handle special kwargs in user-written Triton kernel calls"


Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented on torch/_dynamo/variables/functions.py:
```diff
@@ -843,12 +844,43 @@ def __init__(self, kernel, kernel_idx, grid, **kwargs):
+                new_configs = copy.deepcopy(self.kernel.configs)
+                for config in new_configs:
+                    config.__dict__.update(special_kwargs)
+                new_key = [self.kernel.arg_names[i] for i in self.kernel.key_idx]
```
you dont need this

oulgen commented on torch/_dynamo/variables/functions.py:
```diff
@@ -843,12 +844,43 @@ def __init__(self, kernel, kernel_idx, grid, **kwargs):
+                    config.__dict__.update(special_kwargs)
+                new_key = [self.kernel.arg_names[i] for i in self.kernel.key_idx]
+                # skip other args of autotune, as we only support these
+                new_kernel = autotune(configs=new_configs, key=new_key)(
```
leave the key empty list

oulgen commented on test/inductor/test_triton_kernels.py:
```diff
@@ -1136,6 +1136,92 @@ def f(x, y, dtype_torch, dtype_triton):
 
+    @requires_cuda
+    @skipIfRocm
+    def test_triton_kernel_special_kwargs_with_autotune(self):
```
parametrize by backend all three options

oulgen commented on test/inductor/test_triton_kernels.py:
```diff
@@ -1136,6 +1136,92 @@ def f(x, y, dtype_torch, dtype_triton):
+
+    @requires_cuda
+    @skipIfRocm
+    def test_triton_kernel_special_kwargs_without_autotune(self):
```
ditto

aakhundov committed:
Update on "Handle special kwargs in user-written Triton kernel calls"


Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov committed:
Update on "Handle special kwargs in user-written Triton kernel calls"


Summary: Special kwargs like `num_warps`, `num_stages`, and `num_ctas` can be passed to the Triton kernel call as kwargs. These kwargs are handled in a special way, not being passed to the underlying kernel function directly. In this PR, we move those special kwargs from `kwargs` of the `TritonKernelVariable` in dynamo to `Autotuner`'s `Config` instances (either already existing or newly created for this purpose). As a result, the special kwargs can be codegened correctly as a part of `Config`, not as direct arguments to the kernel `.run`.

Test Plan:

```
python test/inductor/test_triton_kernels.py -k test_triton_kernel_special_kwargs
...
----------------------------------------------------------------------
Ran 6 tests in 6.783s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
